### ISSUE TITLE
passing `any` to short-circuit and/or (`||`/`&&`) loses more precision than necessary

### ISSUE THREAD

**Issue by cakoose** [microsoft/TypeScript#34986]
[TypeScript playground link](http://www.typescriptlang.org/play/?ssl=11&ssc=2&pln=1&pc=1#code/GYVwdgxgLglg9mABMAjACjALkWEBbAIwFMAnAGkQENtKwBPASmwDc4YATRAbwFgAoRIggIAzlEQEU2AnDgAbIrUQBeKogA+6nAG5BAej2IAQgEEAItjBxEpEnBKISRAA72oRdroDkkr4jwwAOYAFuKBROKUIiJBYB5quISk-ILCYGISAEzSsgpKqkiaVLqIBsbmlta29o4ubh7eBJl+ASFhEVTRsfGUOPjEJClCouIEAMw58opIqr0AZHM6+oYA4gDyaxaIACp0zkSIXokDfjAiOHCRXYFglAQKiFDWUHsHPrnTXkNpGQQALJM8jMcIgFsVluUtlYbCQ7A4nK4SO5PId-i0gqFEOErjEbj0+klBnwAL78fhlACylDoxCEcjgIlIj2sdkuQkoIEZmH4oEgsAQyEyaGYln6TK0tEYLDYnF4AmG6VGgOmKkQzBKZVMZgoAElHsEYGAANYAfn4xKAA)

```typescript
function f1(n: number, a: any): void {
  const b1: boolean = a || n;  // BAD: no error reported; 'b1' might get assigned a number
  const b2: boolean = n || a;  // BAD: no error reported; 'b2' might get assigned a number
  const b3: boolean = a && n;  // GOOD: Type 'number' is not assignable to type 'boolean'
  const b4: boolean = n && a;  // BAD: no error reported; 'b4' might get assigned a number
}

// Maybe closer to root cause:
function f2(v: number | any): void {
  const b: boolean = v;  // BAD, I think?
}
```

**Comment by fatcerberus** [comment]
`T | any` and `T & any` are always `any`, regardless of `T`.  This is by design.  Long story short - if you want type safety, you need to get rid of the `any`s posthaste.

**Comment by cakoose** [comment]
Is the reasoning around this design described anywhere?  For example, in an existing issue?

**Comment by fatcerberus** [comment]
Simply because that’s what `any` *does*, it disables type checking in contexts where it occurs.  It’s not really “the widest possible type” like intuition tends to suggest—it’s more of a special “don’t type-check this” flag that you can use in place of a type.

Here’s an issue about it: https://github.com/microsoft/TypeScript/issues/18568

In this particular case, you have an expression of the form `x || y`.  The compiler needs a type for that expression so it constructs a union `TypeOfX | TypeOfY`.  If either one is `any`, the whole type becomes `any`.

**Comment by cakoose** [comment]
I realize that that's what TypeScript is doing and it's not an unreasonable choice.  But I ran into this problem in real-world code and was surprised by it, so I'm just wondering if it's worth doing better.

>  It’s not really “the widest possible type” like intuition tends to suggest—it’s more of a special “don’t type-check this” flag that you can use in place of a type.

Yes, that's exactly the reason this is an issue.  If it were simply the widest possible type (say, `top`), then treating `top | number` like `top` is be fine because we'd eventually catch the error when trying to assign that to `boolean`.

Thanks for the link to the issue!  One difference there is that the issue there is for auto-complete, not for type checking itself.  (It *would* have affected type checking if it were on the return type, though.)

A comment on that issue suggests that there's an important optimization that depends on having `any` subsume the whole union.  On the face of it, it seems like tracking `string | number` is just as expensive as tracking `any | number`, but I don't know anything about the TypeScript compiler code.

**Comment by fatcerberus** [comment]
If you want an actual top type, use `unknown` instead of `any`.

**Comment by cakoose** [comment]
No, I don't want `top`.  I was just saying that treating `any | T` as `any` would be 100% ok *if it were* "the widest possible type".

But since it actually means "don't type-check this", collapsing `any | T` to `any` loses some precision in real-world code in a way that surprised me.
